panic: runtime error: invalid memory address or nil pointer dereference

This panic happens when call w.Close() in a tcp dns server.

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x73a807]

goroutine 378753 [running]:
github.com/miekg/dns.(*Server).readTCP(0xc82146fc70, 0x0, 0x1dcd65000, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/fastcat/src/github.com/miekg/dns/server.go:562 +0x5d7
github.com/miekg/dns.(*defaultReader).ReadTCP(0xc82017e6d8, 0x0, 0x1dcd65000, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/fastcat/src/github.com/miekg/dns/server.go:226 +0x5c
github.com/miekg/dns.(*Server).serve(0xc82146fc70, 0x7f9d658d9a48, 0xc8214a74d0, 0x7f9d658d9a20, 0xc82017fa68, 0xc82220caa0, 0xc, 0xc, 0x0, 0x0, ...)
        /root/fastcat/src/github.com/miekg/dns/server.go:552 +0x450
created by github.com/miekg/dns.(*Server).serveTCP
        /root/fastcat/src/github.com/miekg/dns/server.go:450 +0x3e9

https://github.com/miekg/dns/blob/master/server.go#L683
https://github.com/miekg/dns/blob/master/server.go#L555

--- a/server.go
+++ b/server.go
@@ -535,6 +535,9 @@ Redo:
 	h.ServeDNS(w, req) // Writes back to the client
 
 Exit:
+	if w.tcp == nil {
+		return
+	}
 	// TODO(miek): make this number configurable?
 	if q > maxTCPQueries { // close socket after this many queries
 		w.Close()

--- a/server_test.go
+++ b/server_test.go
@@ -2,6 +2,7 @@ package dns
 
 import (
 	"fmt"
+	"io"
 	"net"
 	"runtime"
 	"sync"
@@ -386,6 +387,46 @@ func TestShutdownTCP(t *testing.T) {
 	err = s.Shutdown()
 	if err != nil {
 		t.Errorf("Could not shutdown test TCP server, %v", err)
+	}
+}
+
+func TestHandlerCloseTCP(t *testing.T) {
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		panic(err)
+	}
+	addr := ln.Addr().String()
+
+	server := &Server{Addr: addr, Net: "tcp", Listener: ln}
+
+	hname := "testhandlerclosetcp."
+	h_triggered := false
+	HandleFunc(hname, func(w ResponseWriter, r *Msg) {
+		h_triggered = true
+		w.Close()
+	})
+	defer HandleRemove(hname)
+
+	go func() {
+		defer server.Shutdown()
+		c := &Client{Net: "tcp"}
+		m := new(Msg).SetQuestion(hname, 1)
+		tries := 0
+	exchange:
+		_, _, err := c.Exchange(m, addr)
+		if err != nil && err != io.EOF {
+			t.Logf("Exchange failed: %s\n", err)
+			if tries == 3 {
+				return
+			}
+			time.Sleep(time.Second / 10)
+			tries += 1
+			goto exchange
+		}
+	}()
+	server.ActivateAndServe()
+	if !h_triggered {
+		t.Fatalf("Handler never called")
 	}
 }
 

